src/coins.cpp

```diff
@@ -577,6 +577,12 @@ void UpdateCoins(const CTransaction &tx, CValidationState &state, CCoinsViewCach
             inputs.SpendCoin(txin.prevout, &txundo.vprevout.back());
         }
     }
+}
+
+void UpdateCoins(const CTransaction &tx, CValidationState &state, CCoinsViewCache &inputs, CTxUndo &txundo, int nHeight)
+{
+    // mark inputs spent
+    SpendCoins(tx, state, inputs, txundo, nHeight);
     // add outputs
```

Comment:
I don't understand why this is part of CTOR? why would we spend coins when updating coins or is this for some other bug fix?

Reply:
seconded

Reply:
This is a refactor with no semantic changes.  UpdateCoins needs to "spend" the inputs of the tx, and add the outputs of the tx into the UTXO set.  But the OTI algorithm needs to do each step separately and in the reverse order (first it AddCoins and second it SpendCoins).  Therefore I took the original code and factored it into a new function called "SpendCoins" and call that function here.